sdk/tests/core/test_az_mqtt5_connection.c

```diff
@@ -434,28 +668,917 @@ static void test_az_mqtt5_connection_open_idle_success(void** state)
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+  _az_mqtt5_connection_reconnecttimeout_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+  _az_mqtt5_connection_reconnecttimeout_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+
+  assert_int_equal(ref_retry_exhausted, 1);
+  assert_int_equal(ref_conn_req, 1);
+  assert_int_equal(ref_conn_rsp_err, 3);
+  assert_int_equal(ref_timeout, 2);
+}
+
+// Testing that a successful connection and disconnection sequence works as expected.
+static void test_az_mqtt5_connection_connect_disconnect_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+  _az_mqtt5_connection_connected_disconnecting(&test_connection);
+}
+
+// Testing that a successful connection, re-connection and disconnection sequence works as expected.
+static void test_az_mqtt5_connection_connect_failure_reconnect_disconnect_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+  _az_mqtt5_connection_reconnecttimeout_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+  _az_mqtt5_connection_connected_disconnecting(&test_connection);
+
+  assert_int_equal(ref_timeout, 1);
+}
+
+// Testing that a successful connection, exhausted re-connection, connection and disconnection
```

Comment:
same question about exhausted to faulted here

Reply:
See above!